src/daemon/db_util.rs

```diff
@@ -0,0 +1,343 @@
+    config: &mut Config,
+    opts: &CliOpts,
+) -> anyhow::Result<(Arc<ManyCar<Arc<RollingDB>>>, Option<Tipset>)> {
+    let mut heaviest_tipset: Option<Tipset> = None;
+    let chain_data_path = chain_path(config);
+    let db_root_dir = db_root(&chain_data_path);
+    let forest_car_db_dir = db_root_dir.join("car_db");
+    if !forest_car_db_dir.is_dir() {
+        fs::create_dir_all(&forest_car_db_dir)?;
+    }
+
+    let mut store = ManyCar::new(Arc::new(open_proxy_db(
+        db_root_dir.clone(),
+        config.db_config().clone(),
+    )?));
+
+    let mut has_errors = false;
+    // Load existing CAR DB(s)
+    for file in WalkDir::new(&forest_car_db_dir)
+        .into_iter()
+        .filter_map(|entry| {
+            if let Ok(entry) = entry {
+                if let Some(filename) = entry.file_name().to_str() {
+                    if filename.ends_with(FOREST_CAR_FILE_EXTENSION) {
+                        return Some(entry.into_path());
+                    }
+                }
+            }
+            None
+        })
+    {
+        match ForestCar::new(RandomAccessFile::open(&file)?) {
+            Ok(car) => {
+                store.read_only(car.into());
+                info!("Loaded car DB at {}", file.display());
+            }
+            Err(err) => {
+                has_errors = true;
+                error!("Error loading car DB at {}: {err}", file.display())
+            }
+        };
+    }
+
+    // Let it check all DB files altogether instead of failing 1 by 1.
+    if has_errors {
+        bail!("Error loading DB file(s).");
+    }
+
+    // TODO: use `--consume-snapshot` CLI option once it's implemented
+    let mut consume_snapshot_file = false;
+    if config.client.snapshot_path.is_none() {
+        let epoch = {
+            if store.read_only_len() == 0 {
+                0
+            } else if let Ok(Some(ts)) = Tipset::load_heaviest(&store, &store) {
+                ts.epoch()
+            } else {
+                0
+            }
+        };
+        fetch_snapshot_if_required(config, epoch, opts.auto_download_snapshot, &db_root_dir)
+            .await?;
+        consume_snapshot_file = true;
+    }
+
+    if !opts.skip_load.unwrap_or_default() {
+        if let Some(path) = &config.client.snapshot_path {
+            let (car_db_path, ts) =
+                import_chain_as_forest_car(path, &forest_car_db_dir, consume_snapshot_file).await?;
+            heaviest_tipset = Some(ts);
+            store.read_only_files(std::iter::once(car_db_path.clone()))?;
+            info!("Loaded car DB at {}", car_db_path.display());
+        }
+    }
+
+    Ok((Arc::new(store), heaviest_tipset))
+}
+
+/// This function validates and stores the CAR binary from `from_path`(either local path or URL) into the `{DB_ROOT}/car_db/`
+/// (automatically trans-code into `.forest.car.zst` format when needed), and returns its final file path and the heaviest tipset.
+async fn import_chain_as_forest_car(
```

Comment:
Let's refactor this function.

- If `from_path` is a URL, download it to `downloaded_car_temp_path`.
- Add a utility function that copies if renaming fails. This should reduce lines 131-163 to two or three lines.
- Hoist out the code that converts from CAR to ForestCAR.zst.

Reply:
Fixed